File: benchmarks/file_io/image_io.m
% image_io.m -- image save/load round trip benchmark, MATLAB side.
%
% Same shape as image_io.qu/.py: 2000x2000 grayscale-from-random image.
% Matty has no imwrite/imread (checked src/graphics.py and builtins.py --
% only a matplotlib-style `imshow` for plotting, no image file codec at
% all) -- this scenario cannot run under Matty; noted in the README rather
% than silently skipped.
%
% Run (cd into this script's own directory first, same reasoning as
% binary_doubles.m):
%   matlab -batch "cd('benchmarks/file_io'); image_io"

W = 2000;
H = 2000;
bmp_path = 'matlab_random.bmp';
png_path = 'matlab_random.png';

rng(5);
m = uint8(round(rand(H, W) * 255));
img = repmat(m, [1 1 3]);

tic;
imwrite(img, bmp_path);
t_save_bmp = toc;
fprintf('imwrite .bmp (%dx%d) : %.4f s\n', W, H, t_save_bmp);

tic;
imwrite(img, png_path);
t_save_png = toc;
fprintf('imwrite .png (%dx%d) : %.4f s\n', W, H, t_save_png);

tic;
img2 = imread(bmp_path);
t_load_bmp = toc;
fprintf('imread .bmp (%dx%d) : %.4f s\n', W, H, t_load_bmp);

[h2, w2, ~] = size(img2);
fprintf('width=%d height=%d\n', w2, h2);
g2 = double(img2(:,:,1));
fprintf('mean pixel (round-tripped) = %.4f\n', mean(g2(:)));
fprintf('mean(original matrix)      = %.4f\n', mean(double(m(:))));

File: benchmarks/file_io/image_io.py
"""image_io.py -- image save/load round trip benchmark, Python (Pillow) side.

Same shape as image_io.qu: 2000x2000 grayscale-from-random image.

Run: python benchmarks/file_io/image_io.py
"""
import time
import numpy as np
from PIL import Image

W = H = 2000
bmp_path = "benchmarks/file_io/py_random.bmp"
png_path = "benchmarks/file_io/py_random.png"

rng = np.random.default_rng(5)
m = (rng.random((H, W)) * 255).round().clip(0, 255).astype(np.uint8)
img = Image.fromarray(m, mode="L").convert("RGB")

t0 = time.perf_counter()
img.save(bmp_path)
t_save_bmp = time.perf_counter() - t0
print(f"PIL save .bmp ({W}x{H}) : {t_save_bmp:.4f} s")

t0 = time.perf_counter()
img.save(png_path)
t_save_png = time.perf_counter() - t0
print(f"PIL save .png ({W}x{H}) : {t_save_png:.4f} s")

t0 = time.perf_counter()
img2 = Image.open(bmp_path)
img2.load()
t_load_bmp = time.perf_counter() - t0
print(f"PIL load .bmp ({W}x{H}) : {t_load_bmp:.4f} s")

print(f"width={img2.width} height={img2.height}")
g2 = np.asarray(img2.convert("L"), dtype=np.float64)
print(f"mean pixel (round-tripped) = {g2.mean():.4f}")
print(f"mean(original matrix)      = {m.astype(np.float64).mean():.4f}")
